### PR TITLE
Promote CRD defaulting to GA

### PR DESCRIPTION
**Pull request by sttts**
/kind api-change

```release-note
CRDs defaulting is promoted to GA. Note: the feature gate CustomResourceDefaulting will be removed in 1.18.
```
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190426-crd-defaulting.md
Feature: https://github.com/kubernetes/enhancements/issues/575

### CODE REVIEW

File: test/e2e/apimachinery/custom_resource_definition.go [hunk 1]
```diff
@@ -251,6 +256,119 @@ var _ = SIGDescribe("CustomResourceDefinition resources [Privileged:ClusterAdmin
...
 	})
 
+	/*
+		Release : v1.17
+		Testname: Custom Resource Definition, defaulting
+		Description: Create a custom resource definition without default. Create CR. Add default and read CR until
+		the default is applied. Create another CR. Remove default, add default for another field and read CR until
+		new field is defaulted, but old default stays.
+	*/
+	framework.ConformanceIt("getting/updating/patching custom resource definition status sub-resource works ", func() {
```
Comment (liggitt) [inline test/e2e/apimachinery/custom_resource_definition.go]: test title doesn't describe testing defaults, is that specifically what this is testing?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go [hunk 2]
```diff
@@ -79,5 +81,5 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	CustomResourceSubresources:      {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	CustomResourceWebhookConversion: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	CustomResourcePublishOpenAPI:    {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
-	CustomResourceDefaulting:        {Default: true, PreRelease: featuregate.Beta},
+	CustomResourceDefaulting:        {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go]: drop check for the feature gate in validation.go#allowDefaults

File: test/e2e/apimachinery/custom_resource_definition.go (hunk 1 above)
Comment (sttts) [reply test/e2e/apimachinery/custom_resource_definition.go]: copy&paste

Comment (sttts) [reply test/e2e/apimachinery/custom_resource_definition.go]: fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go]: done

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go]: note that I only removed the gate check. The `allowDefaults` logic is still needed as it drives the default validation for non-structural, non-pruning CRDs.

**Comment by liggitt** [comment]
> note that I only removed the gate check. The allowDefaults logic is still needed as it drives the default validation for non-structural, non-pruning CRDs.

that matches my expectations, thanks

File: test/e2e/apimachinery/custom_resource_definition.go [hunk 3]
```diff
@@ -251,6 +256,119 @@ var _ = SIGDescribe("CustomResourceDefinition resources [Privileged:ClusterAdmin
...
 	})
 
+	/*
+		Release : v1.17
+		Testname: Custom Resource Definition, defaulting
+		Description: Create a custom resource definition without default. Create CR. Add default and read CR until
+		the default is applied. Create another CR. Remove default, add default for another field and read CR until
+		new field is defaulted, but old default stays.
+	*/
+	framework.ConformanceIt("custom resource defaulting for requests and from storage works ", func() {
```
Comment (liggitt) [inline test/e2e/apimachinery/custom_resource_definition.go]: start this as a regular non-conformance e2e, then promote before freeze once we demonstrate it is not flaky

File: pkg/features/kube_features.go [hunk 4]
```diff
@@ -592,7 +592,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	apiextensionsfeatures.CustomResourceSubresources:      {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	apiextensionsfeatures.CustomResourceWebhookConversion: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	apiextensionsfeatures.CustomResourcePublishOpenAPI:    {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
-	apiextensionsfeatures.CustomResourceDefaulting:        {Default: true, PreRelease: featuregate.Beta},
+	apiextensionsfeatures.CustomResourceDefaulting:        {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Comment (liggitt) [inline pkg/features/kube_features.go]: add a "TODO: remove in 1.18" comment, and a release note that the feature gate will be removed in 1.18

**Comment by sttts** [comment]
Weird `Go toolchain does not support cgo` error in pull-kubernetes-conformance-image-test in `cmd/kubelet`, i.e. totally unrelated it seems.

/retest

**Comment by sttts** [comment]
According to @dims the pull-kubernetes-conformance-image-test job is optional.

/assign @liggitt

**Comment by sttts** [comment]
ptal @liggitt

File: pkg/features/kube_features.go (hunk 4 above)
Comment (sttts) [reply pkg/features/kube_features.go]: done

File: test/e2e/apimachinery/custom_resource_definition.go (hunk 3 above)
Comment (sttts) [reply test/e2e/apimachinery/custom_resource_definition.go]: Conformance follow-up https://github.com/kubernetes/kubernetes/pull/84861

**Pull request merged**